boot/bootutil/include/bootutil/fault_injection_hardening.h

```diff
@@ -0,0 +1,343 @@
+}
+__attribute__((always_inline)) inline
+int fih_not_eq(fih_int x, fih_int y)
+{
+    fih_int_validate(x);
+    fih_int_validate(y);
+    return (x.val != y.val) && fih_delay() && (x.msk != y.msk);
+}
+#else
+/* NOOP */
+__attribute__((always_inline)) inline
+void fih_int_validate(fih_int x)
+{
+    (void) x;
+    return;
+}
+/* NOOP */
+__attribute__((always_inline)) inline
+int fih_int_decode(fih_int x)
+{
+    return x;
+}
+/* NOOP */
+__attribute__((always_inline)) inline
+fih_int fih_int_encode(int x)
+{
+    return x;
+}
+
+__attribute__((always_inline)) inline
+int fih_eq(fih_int x, fih_int y)
+{
+    return x == y;
+}
+__attribute__((always_inline)) inline
+int fih_not_eq(fih_int x, fih_int y)
+{
+    return x != y;
+}
+#endif /* FIH_ENABLE_DOUBLE_VARS */
+
+/* C has a common return pattern where 0 is a correct value and all others are
+ * errors. This function converts 0 to FIH_SUCCESS and any other number to a
+ * value that is not FIH_SUCCESS
+ */
+__attribute__((always_inline)) inline
+fih_int fih_int_encode_zero_equality(int x)
+{
+    if (x) {
+        return FIH_FAILURE;
+    } else {
+        return FIH_SUCCESS;
+    }
+}
+
+#ifdef FIH_ENABLE_CFI
+extern fih_int _fih_cfi_ctr;
+#endif /* FIH_ENABLE_CFI */
+
+fih_int fih_cfi_get_and_increment(void);
+void fih_cfi_validate(fih_int saved);
+void fih_cfi_decrement(void);
+
+/* Label for interacting with FI testing tool. Can be parsed from the elf file
+ * after compilation. Does not require debug symbols.
+ */
+#define FIH_LABEL(str) __asm volatile ("FIH_LABEL_" str "_%=:" ::);
+
+/* Main FIH calling macro. return variable is second argument. Does some setup
+ * before and validation afterwards. Inserts labels for use with testing script.
+ *
+ * First perform the precall step - this gets the current value of the CFI
+ * counter and saves it to a local variable, and then increments the counter.
+ *
+ * Then set the return variable to FIH_FAILURE as a base case.
+ *
+ * Then perform the function call. As part of the funtion FIH_RET must be called
+ * which will decrement the counter.
+ *
+ * The postcall step gets the value of the counter and compares it to the
+ * previously saved value. If this is equal then the function call and all child
+ * function calls were performed.
+ */
+#define FIH_CALL(f, ret, ...) \
+    do { \
+        FIH_LABEL("FIH_CALL_START"); \
+        FIH_CFI_PRECALL_BLOCK; \
+        ret = FIH_FAILURE; \
+        if (fih_delay()) { \
+            ret = f(__VA_ARGS__); \
+        } \
+        FIH_CFI_POSTCALL_BLOCK; \
+        FIH_LABEL("FIH_CALL_END"); \
+    } while (0)
+
+/* FIH return changes the state of the internal state machine. If you do a
+ * FIH_CALL then you need to do a FIH_RET else the state machine will detect
+ * tampering and panic.
+ */
+#define FIH_RET(ret) \
+    do { \
+        FIH_CFI_PRERET; \
+        return ret; \
+    } while (0)
+
+/* Macro wrappers for functions - Even when the functions have zero body this
```

Comment:
move macro doc right above the macro (after `ifdef FIH_ENABLE_CFI`)

Reply:
ack